Src/FluentAssertions/Common/TypeExtensions.cs

```diff
@@ -194,46 +197,62 @@ public static FieldInfo FindField(this Type type, string fieldName, Type preferr
                 : properties.SingleOrDefault();
         }
 
-        public static IEnumerable<MemberInfo> GetNonPrivateMembers(this Type typeToReflect)
+        public static IEnumerable<MemberInfo> GetNonPrivateMembers(this Type typeToReflect, MemberVisibility visibility)
         {
             return
-                GetNonPrivateProperties(typeToReflect)
-                    .Concat<MemberInfo>(GetNonPrivateFields(typeToReflect))
+                GetNonPrivateProperties(typeToReflect, visibility)
+                    .Concat<MemberInfo>(GetNonPrivateFields(typeToReflect, visibility))
                     .ToArray();
         }
 
-        public static IEnumerable<PropertyInfo> GetNonPrivateProperties(this Type typeToReflect,
-            IEnumerable<string> filter = null)
+        public static IEnumerable<PropertyInfo> GetNonPrivateProperties(this Type typeToReflect, MemberVisibility visibility)
         {
             IEnumerable<PropertyInfo> query =
-                from propertyInfo in GetPropertiesFromHierarchy(typeToReflect)
+                from propertyInfo in GetPropertiesFromHierarchy(typeToReflect, visibility)
                 where HasNonPrivateGetter(propertyInfo)
                 where !propertyInfo.IsIndexer()
-                where filter is null || filter.Contains(propertyInfo.Name)
                 select propertyInfo;
 
             return query.ToArray();
         }
 
-        public static IEnumerable<FieldInfo> GetNonPrivateFields(this Type typeToReflect)
+        private static IEnumerable<PropertyInfo> GetPropertiesFromHierarchy(Type typeToReflect, MemberVisibility memberVisibility)
+        {
+            bool includeInternals = memberVisibility.HasFlag(MemberVisibility.Internal);
+
+            return GetMembersFromHierarchy(typeToReflect, type =>
+            {
+                return type
+                    .GetProperties(AllInstanceMembersFlag)
+                    .Where(property => property.GetMethod?.IsPrivate == false)
+                    .Where(property => includeInternals || property.GetMethod?.IsAssembly == false)
```

Comment:
How does this handle properties marked as `private protected` or `protected internal`?

Reply:
AFAIK `private protected` is not relevant here, but `protect internal` was. I've added extra tests to cover this scenario.

Reply:
I'm wondering a bit about what that means to protected members.

We ignore `protected` members
unless it's `protected internal`
unless it `private protected`?